doc/ConnectionPooling.md

```diff
@@ -0,0 +1,387 @@
+### Multiple Connection Pools
+
+Snowflake .NET Driver v4.0.0 provides multiple pools with couple of additional features in comparison to the previous implementation.
+Each pool is identified by the entire <b>connection string</b>. Order of connection string parameters is significant and the same connection parameters
+ordered differently lead to two different pools being used.
+
+All the pool parameters can be controlled from the connection string.
+
+Pool interface is also maintained by the [SnowflakeDbConnectionPool.cs](https://github.com/snowflakedb/snowflake-connector-net/blob/master/Snowflake.Data/Client/SnowflakeDbConnectionPool.cs).
+However, some operations (eg. setting pool parameters from this SnowflakeDbConnectionPool class) are not possible having in mind multiple pools and possibly their different setup.
+For that a [SnowflakeDbSessionPool.cs](/Snowflake.Data/Client/SnowflakeDbSessionPool.cs) is provided by
+- [SnowflakeDbSessionPool.GetPool(connectionString)](https://github.com/snowflakedb/snowflake-connector-net/blob/master/Snowflake.Data/Client/SnowflakeDbConnectionPool.cs#L45)
+- [SnowflakeDbSessionPool.GetPool(connectionString, securePassword)](https://github.com/snowflakedb/snowflake-connector-net/blob/master/Snowflake.Data/Client/SnowflakeDbConnectionPool.cs#L51).
+to control pool settings from the code. Changed pool settings are not reflected by their connection string therefore recommended way is to control the pool from the connection string.
+
+### Pool Lifecycle
+
+Single pool is instantiated each time an application creates and opens a connection for the first time using particular connection string.
+From that moment the pool tracks and maintains connections matching exactly this connection string.
+Pool is responsible for destroying and recreating connections which are old enough (see [Expiration Timeout](#expiration-timeout)).
+Number of connections is maintained within [Minimum pool size](#minimum-pool-size) and [Maximum pool size](#maximum-pool-size).
+Connections in all their statuses are tracked:
+- opening phase
+- busy phase
+- closed and returned to the pool (idle)
+User can clean up the pool using methods: [Clear Pool](#clear-pool).
+
+### Connection Lifecycle
+
+#### Opening
+
+Before connection is opened it is `scheduled` to be opened in the pool on a waiting queue. Connections scheduled to open are counted to the pool size.
+After connection is successfully opened it is counted as a `busy` in a separate list. Pool can provide already opened `idle` connection or if there are none it will schedule new one to open.
+When [Maximum pool size](#maximum-pool-size) is reached connection is waiting to be opened within period of time controlled with [Pool Size Exceeded Timeout](#pool-size-exceeded-timeout).
+When the timeout is exceeded then an exception will get thrown.
+
+#### Busy
+
+`Busy` connection is provided by the pool but it is counted to the pool size. It is returned to be reused during Close operation.
```

Comment:
I would write "and is counted to the pool size", not "but".
The second sentence is actually not about being "busy", because you can get a busy session by creating a new one as well.